### PR TITLE
Add support for context manager dependencies

### LINKED ISSUE #24: Add support for context manager dependencies
There is currently support for dependency-generators, with yield syntax. It is also worth adding support for context managers (`__enter__`, `__exit__`, `__aenter__` and `__aexit__`), as they implement the same functionality, but are more convenient to use outside DI.

### PR DESCRIPTION
**Pull request by cofob**
Changes:
- Fixed one typo
- Added a test for annotated gen dependency
- Added support and tests for context manager dependencies

Also a bit offtopic, but I think flake8 has gone overboard with the number of checks. Almost every line is a noqa.

Closes #24

### CODE REVIEW

**Review by s3rius — APPROVED** [review]
Excellent work. Thank you a lot. 

This is  a very valuable and needed contribution. I haven't thought of it.

**Comment by s3rius** [comment]
Regarding the flake8, we might transfer checks to ruff as we've done in other repos. It'd be a bit more relaxed then flake8, so it should reduce number of noqas. But thanks for noticing.

File: tests/test_graph.py [hunk 1]
```diff
@@ -111,6 +112,76 @@ def testfunc(a: int = Depends(dep1)) -> int:
...
+        nonlocal starts  # noqa: WPS420
+        nonlocal closes  # noqa: WPS420
+
+        await asyncio.sleep(0.001)
+        starts += 1
+
+        try:
+            yield 1
+        finally:
+            await asyncio.sleep(0.001)
```
[the remark this answers is not in the document]
Comment (cofob) [reply tests/test_graph.py:168]: I don't know why it's there, I just copied the tests for generators and changed them for context managers.

File: tests/test_graph.py [hunk 2]
```diff
@@ -111,6 +112,76 @@ def testfunc(a: int = Depends(dep1)) -> int:
...
+            closes += 1
+
+    def testfunc(a: int = Depends(dep1)) -> int:
+        return a
+
+    with DependencyGraph(testfunc).sync_ctx({}) as sctx:
+        with pytest.raises(RuntimeError):
+            assert sctx.resolve_kwargs() == {"a": 1}
```
[the remark this answers is not in the document]
Comment (cofob) [reply tests/test_graph.py:176]: I think a check on the return value is still needed here. And again, the same check is in the tests for generators, I just copied them.

File: tests/test_graph.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (s3rius) [reply tests/test_graph.py:168]: Can you remove it in both places?

[the remark this answers is not in the document]
Comment (cofob) [reply tests/test_graph.py]: Fixed in 6b85e2bae2e866bb1532e9ec7c12642b97082f21

**Comment by s3rius** [comment]
Thank you very much.

**Pull request merged**